### PR TITLE
[FIX] delivery: Prevent delivery product to be set as sellable

### PR DESCRIPTION
**Pull request by InCeyN**
Currently, if a product is one configured on a delivery method - it is
set to be 'is_delivery' = true.
This has it's limitations if they then configure this same product as
being sell-able or purchasable. Once they have done so and sell/purchase
a product, the invoice status in the process does not follow the
standard behavior.

The point of this commit is to avoid running into this issue and
preventing the product to be configured as sell-able or purchasable if
'is_delivery' is true.

Description of the issue/feature this PR addresses:
opw-2829439

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/delivery/models/product_template.py [hunk 1]
```diff
@@ -11,3 +12,11 @@ class ProductTemplate(models.Model):
         string="HS Code",
         help="Standardized code for international shipping and goods declaration. At the moment, only used for the FedEx shipping provider.",
     )
+
+    def write(self, values):
```
Comment (fmdl) [inline addons/delivery/models/product_template.py:16]: maybe add sudo if user have not access to this method.
maybe check for all delivery.carrier
maybe put this in a @api.contraints
maybe check for all products (1 sql query)

```python
   def write(self, vals):
        if vals.get('sale_ok') is False or vals.get('purchase_ok') is False:
                carriers = self.env['delivery.carrier'].sudo().\
                    search([('product_id', 'in', self.product_variant_ids.ids)])
                if carriers:
                    raise UserError(_('This product is set as delivery product of thoses shipping methods:\n%s', '\n'.join(carriers.mapped('display_name')))
        return super().write(vals)
```

Comment (InCeyN) [reply addons/delivery/models/product_template.py]: Thanks for your input, I did indeed take your advice into consideration.
Kr

File: addons/delivery/models/product_template.py [hunk 2]
```diff
@@ -11,3 +12,13 @@ class ProductTemplate(models.Model):
         string="HS Code",
         help="Standardized code for international shipping and goods declaration. At the moment, only used for the FedEx shipping provider.",
     )
+
+    @api.constrains('sale_ok', 'purchase_ok')
+    def _check_is_delivery(self):
+        for record in self:
+            if record.sale_ok or record.purchase_ok:
+                carriers = self.env['delivery.carrier'].sudo().\
+                    search([('product_id', 'in', record.product_variant_ids.ids)])
```
Comment (Whenrow) [inline addons/delivery/models/product_template.py:21]: I would search all delivery products before the loop to make only one db request

**Comment by alt-odoo** [comment]
@Whenrow @amoyaux 

Can we merge this if OK for you please?
Thanks!

**Comment by wtaferner** [comment]
Hmm, a constraint means that if we want to use the product in an invoice only we will not be able to use it...not even temporarily...what about a warning instead?

**Comment by alt-odoo** [comment]
The idea is indeed to not use such products in invoices, there is no real need of this.

**Pull request closed without merging**